src/base/QXmppHttpUploadIq.cpp

```diff
@@ -0,0 +1,262 @@
+{
+    return d->size;
+}
+
+/// Sets the file's size in bytes.
+
+void QXmppHttpUploadRequestIq::setSize(const qint64 &size)
+{
+    d->size = size;
+}
+
+/// Returns the (optional) MIME-type of the file.
+
+QMimeType QXmppHttpUploadRequestIq::contentType() const
+{
+    return d->contentType;
+}
+
+/// Sets the MIME-type of the file. This is optional.
+
+void QXmppHttpUploadRequestIq::setContentType(const QMimeType &type)
+{
+    d->contentType = type;
+}
+
+bool QXmppHttpUploadRequestIq::isHttpUploadRequestIq(const QDomElement &element)
+{
+    if (element.tagName() == "iq") {
+        QDomElement request = element.firstChildElement("request");
+        return !request.isNull() && request.namespaceURI() == ns_http_upload;
+    }
+    return false;
+}
+
+/// \cond
+void QXmppHttpUploadRequestIq::parseElementFromChild(const QDomElement &element)
+{
+    QDomElement request = element.firstChildElement("request");
+    d->fileName = request.attribute("filename");
+    d->size = request.attribute("size").toLongLong();
+    if (request.hasAttribute("content-type")) {
+        QMimeDatabase mimeDb;
+        QMimeType type = mimeDb.mimeTypeForName(request.attribute("content-type"));
+        if (!type.isDefault() && type.isValid())
+            d->contentType = type;
+    }
+}
+
+void QXmppHttpUploadRequestIq::toXmlElementFromChild(QXmlStreamWriter *writer) const
+{
+    writer->writeStartElement("request");
+    writer->writeAttribute("xmlns", ns_http_upload);
+    // filename and size are required
+    writer->writeAttribute("filename", d->fileName);
+    writer->writeAttribute("size", QString::number(d->size));
+    // content-type is optional
+    if (!d->contentType.isDefault() && d->contentType.isValid())
+        writer->writeAttribute("content-type", d->contentType.name());
+    writer->writeEndElement();
+}
+/// \endcond
+
+class QXmppHttpUploadSlotIqPrivate
+{
+public:
+    QUrl putUrl;
+    QUrl getUrl;
+    QMap<QString, QString> headerFields;
+};
+
+QXmppHttpUploadSlotIq::QXmppHttpUploadSlotIq()
+    : d(new QXmppHttpUploadSlotIqPrivate())
+{
+}
+
+QXmppHttpUploadSlotIq::~QXmppHttpUploadSlotIq()
+{
+    delete d;
+}
+
+/// Returns the URL for uploading via. HTTP PUT.
+
+QUrl QXmppHttpUploadSlotIq::putUrl() const
+{
+    return d->putUrl;
+}
+
+/// Sets the URL the client should use for uploading.
+
+void QXmppHttpUploadSlotIq::setPutUrl(const QUrl &putUrl)
+{
+    d->putUrl = putUrl;
+}
+
+/// Returns the URL to where the file will be served.
+
+QUrl QXmppHttpUploadSlotIq::getUrl() const
+{
+    return d->getUrl;
+}
+
+/// Sets the download URL.
+
+void QXmppHttpUploadSlotIq::setGetUrl(const QUrl &getUrl)
+{
+    d->getUrl = getUrl;
+}
+
+/// Returns a map of header fields (header name -> content) that need to be
+/// included in the PUT (upload) request. This won't contain any other fields
+/// than: "Authorization", "Cookie" or "Expires".
+
+QMap<QString, QString> QXmppHttpUploadSlotIq::headerFields() const
```

Comment:
Would putHeaderFields clarify that these headers only apply to PUT?

Reply:
Let's call it just `putHeaders` then.

Reply:
Even better